hpx/util/checkpoint.hpp

```diff
@@ -0,0 +1,516 @@
+    ///////////////////////////////////
+    /// Operator<< Overload
+    ///
+    /// \param ost           Output stream to write to.
+    ///
+    /// \param ckp           Checkpoint to copy from.
+    ///
+    /// This overload is the main way to write data from a
+    /// checkpoint to an object such as a file. Inside
+    /// the function, the size of the checkpoint will
+    /// be written to the stream before the checkpoint's
+    /// data. The operator>> overload uses this to read
+    /// the correct number of bytes. Be mindful of this
+    /// additional write and read when you use different
+    /// facilities to write out or read in data to a
+    /// checkpoint!
+    ///
+    /// \returns Operator<< returns the ostream object.
+    ///
+    std::ostream& operator<<(std::ostream& ost, checkpoint const& ckp)
+    {
+        // Write the size of the checkpoint to the file
+        int64_t size = ckp.size();
+        ost.write(reinterpret_cast<char const*>(&size), sizeof(int64_t));
+        // Write the file to the stream
+        ost.write(ckp.data.data(), ckp.size());
+        return ost;
+    }
+    ///////////////////////////////////
+    /// Operator>> Overload
+    ///
+    /// \param ist           Input stream to write from.
+    ///
+    /// \param ckp           Checkpoint to write to.
+    ///
+    /// This overload is the main way to read in data from an
+    /// object such as a file to a checkpoint.
+    /// It is important to note that inside
+    /// the function, the first variable to be read is
+    /// the size of the checkpoint.
+    /// This size variable is written to
+    /// the stream before the checkpoint's
+    /// data in the operator<< overload.
+    /// Be mindful of this
+    /// additional read and write when you use different
+    /// facilities to read in or write out data from a
+    /// checkpoint!
+    ///
+    /// \returns Operator>> returns the ostream object.
+    ///
+    std::istream& operator>>(std::istream& ist, checkpoint& ckp)
+    {
+        // Read in the size of the next checkpoint
+        int64_t length;
+        ist.read(reinterpret_cast<char*>(&length), sizeof(int64_t));
+        ckp.data.resize(length);
+        // Read in the next checkpoint
+        ist.read(ckp.data.data(), length);
+        return ist;
+    }
+
+    //Function object for save_checkpoint
+    namespace detail {
+        struct save_funct_obj
+        {
+            template <typename... Ts>
+            checkpoint operator()(checkpoint&& c, Ts&&... ts) const
+            {
+                //Create serialization archive from checkpoint data member
+                hpx::serialization::output_archive ar(c.data);
+                //Serialize data
+                int const sequencer[] = {//Trick to expand the variable pack
+                    (ar << ts,
+                        0)...};      //Takes advantage of the comma operator
+                (void) sequencer;    // Suppress unused param. warnings
```

Comment:
For MSVC you'll have to write this as:
```
int const sequencer[] = {
    0, (ar << ts, 0)...
};
```
this avoids generating a possibly zero-sized array this compiler is complaining about.

Reply:
Will fix!